src/cpp/client/channel_cc.cc

```diff
@@ -35,17 +35,82 @@
 #include <grpc/slice.h>
 #include <grpc/support/alloc.h>
 #include <grpc/support/log.h>
+#include <grpc/support/thd.h>
 #include "src/core/lib/profiling/timers.h"
 
 namespace grpc {
 
+namespace {
+void WatchStateChange(void* arg);
+}  // namespace
+
+// Constantly watches channel connectivity status to reconnect a transiently
+// disconnected channel. This is a temporary work-around before we have retry
+// support.
+class ChannelConnectivityWatcher {
+ public:
+  explicit ChannelConnectivityWatcher(Channel* channel)
+      : channel_(channel), thd_id_(0), shutting_down_(0) {}
+
+  void WatchStateChangeImpl() {
+    grpc_connectivity_state state = GRPC_CHANNEL_IDLE;
+    while (state != GRPC_CHANNEL_SHUTDOWN) {
+      channel_->WaitForStateChange(state, gpr_inf_future(GPR_CLOCK_REALTIME));
+      if (gpr_atm_no_barrier_load(&shutting_down_) == 1) {
+        break;
+      }
+      state = channel_->GetState(false);
+    }
+  }
+
+  void StartWatching() {
+    const char* disabled_str =
+        std::getenv("GRPC_DISABLE_CHANNEL_CONNECTIVITY_WATCHER");
```

Comment:
Do we want to use [`gpr_getenv`](https://github.com/grpc/grpc/blob/6331bcf12f0690d7dc07de5d41515985506c5058/src/core/lib/support/env.h#L34) instead?